### DISCUSSION TITLE
YAML Device Component Import

### DISCUSSION THREAD

**Discussion by moiolilu** [netbox-community/netbox#12183] [Q&A]
When exporting a device type via the All data option, it allows you to export all the information about the device type via a file, including the components attached (interfaces, rear-ports, front-ports, etc.).  Then, if importing that YAML file to another Netbox instance correctly creates the device type along with the component templates as well.  

Is there a way to import a YAML containing Device information INCLUDING interface/port information? I have tried to format a YAML file to achieve this, which I based off the device type YAML, however when the import finishes the device is created without any of its components that were provided in the file.  

Any help is appreciated!

**Comment by candlerb** [comment]
I don't think so.

I note that for Devices, the Export option is "All Data (CSV)", not "Add Data (YAML)". In the [source code](https://github.com/netbox-community/netbox/blob/develop/netbox/dcim/models/devices.py), DeviceType and ModuleType have `to_yaml` methods giving all the linked components; Device does not.

Similarly, in the relevant [import code](https://github.com/netbox-community/netbox/blob/develop/netbox/dcim/views.py), there are related_object_forms for DeviceTypeImportView and ModuleTypeImportView, but not for DeviceBulkImportView.

I'm not sure how useful the facility would be to import a device with its components in one step, which is probably why it wasn't implemented. I can also see complexities around what you'd do with virtual chassis, where the master device inherits all the interfaces of the rest of the VC.

In principle, I think it *would* be useful to be able to import Device + interfaces + IP Addresses (because this would also allow you to import the Device's Primary IP address, which is not currently possible) - but that would add a lot more implementation complexity.

What I *can* offer is a script, where after you import a Device, you can run the script to create all the missing components - it copies them from the corresponding DeviceType:
https://github.com/netbox-community/customizations/blob/master/scripts/add_device_type_components.py
This will only give you the interfaces in the DeviceType, not any additional ones like loopbacks or subinterfaces - you'll still have to import those separately.
